src/PathFinding.h

```diff
@@ -0,0 +1,91 @@
+#include <vector>
+#include <map>
+#include <functional>
+#include <algorithm>
+#include <limits>
+
+#ifndef PATH_FINDING_H
+#define PATH_FINDING_H
+
+template<typename Node, typename Key>
+std::vector<Node*> find_shortest_path(
+    Node& start,
+    std::function<Key (Node&)> get_key,
+    std::function<std::vector<Node*> (Node&)> get_neighbors,
+    std::function<bool (Node&)> is_goal
+)
+{
+    std::vector<Node*> to_visit;
+    std::map<Key, bool> visited;
+    std::map<Key, float> distance;
+    std::map<Key, Node*> previous;
+
+    to_visit.push_back(&start);
+    distance[get_key(start)] = 0;
+
+    Node* last = nullptr;
+
+    // Convenience distance accessor
+    auto get_distance = [&](Node& node) -> float {
+        Key key = get_key(node);
+        return distance.find(key) != distance.end()
+            ? distance[key]
+            : std::numeric_limits<float>::infinity();
+    };
+
+    while (to_visit.empty() == false) {
```

Comment:
` == false` is funny. Why not `!` ?

Reply:
😂  probably a habit from the JS world and also `!` is so slim and easy to miss